borsh-derive-internal/src/enum_ser.rs

```diff
@@ -109,3 +80,196 @@ pub fn enum_ser(input: &ItemEnum, cratename: Ident) -> syn::Result<TokenStream2>
+}
+fn named_fields(
+    cratename: &Ident,
+    enum_ident: &Ident,
+    variant_ident: &Ident,
+    discriminant_value: &TokenStream2,
+    fields: &FieldsNamed,
+) -> syn::Result<VariantParts> {
+    let mut where_predicates: Vec<WherePredicate> = vec![];
+    let mut variant_header = TokenStream2::new();
+    let mut variant_body = TokenStream2::new();
+    let mut dot_dot_pattern_match: Option<syn::Token![..]> = None;
+    for field in &fields.named {
+        let field_name = if contains_skip(&field.attrs) {
+            dot_dot_pattern_match = Some(syn::Token![..](Span::call_site()));
+            None
+        } else {
+            Some(field.ident.clone().unwrap())
+        };
+        if let Some(ref field_name) = field_name {
+            variant_header.extend(quote! { #field_name, });
+        }
+        if !contains_skip(&field.attrs) {
+            let field_type = &field.ty;
+            where_predicates.push(
+                syn::parse2(quote! {
+                    #field_type: #cratename::ser::BorshSerialize
+                })
+                .unwrap(),
+            );
+            variant_body.extend(quote! {
+                 #cratename::BorshSerialize::serialize(#field_name, writer)?;
+            })
+        }
+    }
+    if let Some(dot_dot) = dot_dot_pattern_match {
+        variant_header.extend(quote! { #dot_dot });
+    }
+    variant_header = quote! { { #variant_header }};
+    let variant_idx_body = quote!(
+        #enum_ident::#variant_ident { .. } => #discriminant_value,
+    );
+    Ok(VariantParts {
+        where_predicates,
+        variant_header,
+        variant_body,
+        variant_idx_body,
+    })
+}
+
+fn unnamed_fields(
+    cratename: &Ident,
+    enum_ident: &Ident,
+    variant_ident: &Ident,
+    discriminant_value: &TokenStream2,
+    fields: &FieldsUnnamed,
+) -> syn::Result<VariantParts> {
+    let mut where_predicates: Vec<WherePredicate> = vec![];
+    let mut variant_header = TokenStream2::new();
+    let mut variant_body = TokenStream2::new();
+    for (field_idx, field) in fields.unnamed.iter().enumerate() {
+        let field_idx = u32::try_from(field_idx).expect("up to 2^32 fields are supported");
+        if contains_skip(&field.attrs) {
+            let field_ident = Ident::new(format!("_id{}", field_idx).as_str(), Span::call_site());
+            variant_header.extend(quote! { #field_ident, });
+            continue;
+        } else {
+            let field_type = &field.ty;
+            where_predicates.push(
+                syn::parse2(quote! {
+                    #field_type: #cratename::ser::BorshSerialize
+                })
+                .unwrap(),
+            );
+
+            let field_ident = Ident::new(format!("id{}", field_idx).as_str(), Span::call_site());
+            variant_header.extend(quote! { #field_ident, });
+            variant_body.extend(quote! {
+                #cratename::BorshSerialize::serialize(#field_ident, writer)?;
+            })
```

Comment:
Let's align the order of blocks with the one in "named_fields":

```suggestion
            let field_ident = Ident::new(format!("id{}", field_idx).as_str(), Span::call_site());
            variant_header.extend(quote! { #field_ident, });
            let field_type = &field.ty;
            where_predicates.push(
                syn::parse2(quote! {
                    #field_type: #cratename::ser::BorshSerialize
                })
                .unwrap(),
            );
            variant_body.extend(quote! {
                #cratename::BorshSerialize::serialize(#field_ident, writer)?;
            })
```